[WIP][KVConnector] Retrive KV events from LMCache

hickeyma opened the pull request:
<!-- markdownlint-disable -->

## Purpose
Updates to the LMCache connector to enable KV events generated by LMCache to be retrieved and published by the scheduler.

This is part of the solution for:
- https://github.com/LMCache/LMCache/issues/1846 to enable KV events to be generated in LMCache
- RFC [[RFC]: KV-Cache Interoperability API Standardization ](https://github.com/vllm-project/vllm/issues/20492) which proposes "formalizing KV Events as the public contract for any component emitting or consuming KV-Cache lifecycle events, including external indexers, routers, and engines". 

**Note to reviewers: Do NOT merge until the following merge first:**
1. Updated code to use event property instead of the stats property. Property added in #28309.
2. https://github.com/LMCache/LMCache/issues/1846 is merged first.

## Test Plan

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

hickeyma committed:
Retrive KV events from LMCache

Updates to the LMCache connector to enable events generated by LMCache
to be retrieved and published by the scheduler.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

markmc commented:
My initial reaction is that if a connector needs to send events from the worker-side to the scheduler-side, we should add events to `KVConnectorOutput` and not abuse the generic nature of `KVConnectorStats`

markmc commented on vllm/distributed/kv_transfer/kv_connector/v1/lmcache_connector.py:
```diff
@@ -26,6 +31,50 @@
+    def reduce(self) -> dict[str, int | float]:
+        return {
+            "kv_events": 0,
+        }
```
Gemini wouldn't be the only one confused by this! It's a nice example of why we should have an explicit channel for events, rather than using the metrics channel

hickeyma commented:
> My initial reaction is that if a connector needs to send events from the worker-side to the scheduler-side, we should add events to KVConnectorOutput and not abuse the generic nature of KVConnectorStats

I agree @markmc, thanks. This is my first effort by piggybacking on the stats. I will push a separate PR which adds events to KVConnectorOutput first and then rebase this accordingly..

hickeyma committed:
Update following kv events added to KVConnectorOutput

This commit follows recommendation from @markmc to use
a specific event property instead of piggybacking on
stats. PR #28309 adds the events property to KVConnectorOutput and
this commit picks up the new property and uses it to pass the events
from worker side to scheduler side.

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
Thanks @markmc for the review and feedback. Pushed PR #28309 for adding events property to `KVConnectorOutput` and also updated this PR accordingly.

hickeyma commented:
Closing as superseded by #28309
